### PR TITLE
Vagrant environment with Docker-compose orchestration

### PR DESCRIPTION
**Pull request by mattma**
Download my [branch](https://github.com/mattma/JSHomes-Platform) in your local machine. Follow the [instruction](https://github.com/mattma/JSHomes-Platform/tree/master/docs) to setup your JSHome wordpress. Here is the [TODO](https://github.com/mattma/JSHomes-Platform/tree/master/docs#todo) for next PR.

### CODE REVIEW

**Comment by mattma** [comment]
@ericelliott Ping for review!

**Comment by ericelliott** [comment]
Thanks. I'll take a look ASAP. Just decided to take most of the weekend off. =)

**Comment by ericelliott** [comment]
I'm not familiar with any of this tech, so calling in reinforcements from Twitter. Do you know any experts who could comment?

**Comment by mattma** [comment]
Sure. I would be happy to modify the code if any expert comment on it. Have you tried to follow the README and run a local instance of wordpress yourself?

**Comment by robmazan** [comment]
Following steps build was successful, but on startup I get:

```
Cannot start container 49859b4f99dbec7b971785b41eb5c104003a7e9da8864fe4ed1e10ed4a4569c4: [8] System error: no such file or directory
```

Some thoughts off the top of my head:

There is an [official image](https://registry.hub.docker.com/_/wordpress/) for WordPress on the Docker HUB. I think it would be easier to use that. Even if you want to customize, you can have it in your FROM.

IMO using something more visual would aid understanding what's going on for newbies. [Kitematic](https://kitematic.com/) is now available for Windows and OS X too. If Kitematic doesn't suit for you, maybe [DockerUI](https://github.com/crosbymichael/dockerui) will.

I would also consider using some `boot2docker` based solution rather than pulling a complete Ubuntu, as it is only used to run Docker. For example [Docker Machine](https://docs.docker.com/machine/) is my weapon of choice these days. As an alternative solution, if you stick with Vagrant, you can use [Docker provisioner](http://docs.vagrantup.com/v2/provisioning/docker.html).

I prefer using [private network](http://docs.vagrantup.com/v2/networking/private_network.html) instead of port forwarding to avoid accidental port collisions. In this case I'm adding the IP of the VM to the hosts file to make it easier to access.

**Comment by ericelliott** [comment]
> There is an official image for WordPress on the Docker HUB. I think it would be easier to use that. Even if you want to customize, you can have it in your FROM.

I was also confused about why you didn't use the official image...

> IMO using something more visual would aid understanding what's going on for newbies. Kitematic is now available for Windows and OS X too. If Kitematic doesn't suit for you, maybe DockerUI will.

Kitematic sounds ideal.


@robmazan would you be willing to put together a minimal Docker config using Kitematic that will run the official WordPress image and let users edit the contents of the `wp-content` folder?

**Comment by ericelliott** [comment]
@robmazan It looks like there's currently [no UI to add custom volumes](https://github.com/kitematic/kitematic/issues/376) that the host OS can mount? How do you work around that?

Getting Host OS access to containers is the part I was banging my head against during [my own failed attempt](https://github.com/jshomes/JSHomes-Platform/blob/master/docs/experimental/docker.md).

**Comment by robmazan** [comment]
> It looks like there's currently no UI to add custom volumes that the host OS can mount? How do you work around that?

@ericelliott Yeah, seems like Kitematic is simply not there yet. Turned out it doesn't support container linking as well...

So as it looks now `docker-machine` or `boot2docker` with `DockerUI` seems to be a good fallback.

Unfortunately I don't have too much time recently, and don't have any experience with Wordpress. But if you give me some info on how this setup is planned to be used, and what do you expect to be a general process of developing with it, I'll try my best to figure out a way to make it work.

**Comment by mattma** [comment]
@robmazan Default wordpress docker image does not give any customization that we want to do. For example, we have a configuration file `.env`, how could the config be picked up by wordpress docker image without our custom hook?

Second, `boot2docker`, boot2docker is a great project, do not get me wrong. I love it. The problem that I used `boot2docker`, it is a great tool to quickly prototype something, run a dockerfile to get an image, but it set Docker daemon on the host machine. It is global. Vagrant is much isolated and fine-grain control. 

Another things about boot2docker, it is based on `darwin`, the OS. Vagrant is based on `linux`. Most likely, the server would run in Linux environment. so it is a true reflection of the production environment. 

Could you provide in which step that you got problem? exact step to reproduce the issue?

My solution works and should get everyone bootstrap the wordpress in seconds. I agrees that Kitmatic and other UI solution could be picked up easily by newcomers. but the bare bone of the logic still going to be in the code. It is just a representation of the bash scripts.

**Comment by mattma** [comment]
@robmazan @ericelliott I just pulled my branch in a fresh state, and follow the README to start the wordpress, and it works as advertised. Confirmed.

In my understanding, once I finished the custom `.env`, it should be all done.

**Comment by robmazan** [comment]
Hi @mattma,

> Default wordpress docker image does not give any customization that we want to do. For example, we have a configuration file .env, how could the config be picked up by wordpress docker image without our custom hook?

That's why I wrote:

> Even if you want to customize, you can have it in your FROM.

I mean using official WordPress as base image, and build your layer above that. That way you can cut the maintenance cost as you only have to deal with your modifications, not the whole stack.

> The problem that I used boot2docker, it is a great tool to quickly prototype something, run a dockerfile to get an image, but it set Docker daemon on the host machine. It is global. Vagrant is much isolated and fine-grain control.

Sorry, I don't get you here. The `boot2docker` was created to make it possible to run Docker on non-Linux based systems. On Linux you can leverage LXC directly, while the same is not possible for example on Windows. That's why you need a virtual linux where you can run docker. Given that what I don't get is what you mean by "it set Docker daemon on the host machine".

> Another things about boot2docker, it is based on darwin, the OS. Vagrant is based on linux.

Boot2docker is based on Tiny Core Linux. Vagrant is a Ruby project originally created to make it easier to setup VirtualBox environments - used primarily to setup virtual Linux instances.

I admit it is easier to use an Ubuntu as it is easier to persist your stuff (`boot2docker` using RAM based FS), and a lot more easier to install additional stuff.

> Could you provide in which step that you got problem? exact step to reproduce the issue?

I'm off for the weekend now, but will get back to you with that next week.

**Comment by robmazan** [comment]
@mattma @ericelliott Ok, seems like I have the root cause of this error:

> Cannot start container 49859b4f99dbec7b971785b41eb5c104003a7e9da8864fe4ed1e10ed4a4569c4: [8] System error: no such file or directory

I was running the build on Windows. Now when I cloned the repo, git converted all text files to CRLF line endings - including `start.sh`. When I was running `docker-compose up -d`, it died at the first line of `start.sh`, as on Linux it looked like:

```sh
set -e^M
```

I think this CRLF conversion can be prevented via `.gitattributes` (haven't tried it now).

Wrt **Kitematic**:

You can use it quite simply with the following trade-offs:

  * No docker-compose support (yet)
  * Project root must be inside of your Users folder on Win (don't know about iOS)
  * To initialize instances you must run two commands directly from Docker CLI of the Kitematic

Just click on Docker CLI and enter:

```sh
docker run -d -e MYSQL_ROOT_PASSWORD=jshomes --name mysql mysql
docker run -d -e WORDPRESS_DB_PASSWORD=jshomes --link mysql:mysql -v /c/Users/path/to/project:/var/www/html -p 80:80 --name wordpress wordpress
```

Where `/c/Users/path/to/project` should be your project directory.

Note how easily you can open a browser pointing at the running WP, or open the volume with WP files created by the install:

image

Again, I'm not familiar with WordPress, so this setup might not work for you, just wanted to show an alternative approach that I find a bit more compact.

**Comment by ericelliott** [comment]
@robmazan

The point of docker-compose is that we will eventually have several services running, including the WordPress host, a Node server, and (probably) two different databases -- one for the WordPress instance, and one for the Node service.

We want to be able to launch the app with a single command. Using your suggested configuration, it looks like we'll have to launch each service individually? That's more developer overhead than we want.

**Comment by robmazan** [comment]
@ericelliott 

You can still batch those commands to a `.sh`, and then you have only one command. Not that pretty, I agree. :)

If you expect your target audience to be able to install Python, then you can even use `docker-compose` on the Kitematic based setup, as that's only a `pip install -U docker-compose` from there.

If you don't want that, then the Vagrant based solution seems to be optimal. Picking that one, I still suggest:

  * Using the official WordPress image (at least as a base) to cut maintenance cost
    * If you make a custom version, that can be pushed to Docker HUB too, so you can cut the build time
  * Using a smaller distro
    * If `boot2docker` is not enough, maybe you want to check [Alpine](https://registry.hub.docker.com/_/alpine/)
  * Using private network instead of port forwarding
  * Plus I think DockerUI is a nice to have and easy to add

Sorry if I seem to be captious, just trying to get your perspective to help you the best of my ability. :)

**Comment by ericelliott** [comment]
> Using the official WordPress image (at least as a base) to cut maintenance cost

I don't understand why this is not an option here. We should be able to use the official container as a base and add our modifications on top of that.

> Plus I think DockerUI is a nice to have and easy to add

I'm not sure what the benefit is without docker-compose support?

**Comment by robmazan** [comment]
> I'm not sure what the benefit is without docker-compose support?

@ericelliott Ok, so as I see now you can go with the Vagrant based solution, or with the Kitematic based one.

If you choose Vagrant based, you have `docker-compose` in your VM, but you don't have any visual aid (to check logs, ports, etc. of the running Docker containers) - that's where Docker UI can help out.

If you pick the Kitematic based version, you don't have `docker-compose` by default, but I think you can install it on the host pretty easily (at least for Mac - I had difficulties on Win, recently overcame those as well though). With this solution, you have Kitematic as your visual aid, so you don't need Docker UI in this case.

**Comment by ericelliott** [comment]
@mattma Is there any way we can [use the official WordPress container as a base](https://docs.docker.com/reference/builder/#from) and modify from that?

**Comment by mattma** [comment]
@ericelliott I will do that. Will work on it next week.

**Comment by ericelliott** [comment]
@mattma Thanks Matt!

**Comment by mattma** [comment]
@ericelliott I hope that the PR could be merged in as soon as it has been fully tested. No one has ever reported that the current one is working or not, at least, it works on my testing VM. I hope that I could find all the bugs and make sure the scripts work as it should.

**Comment by ericelliott** [comment]
> I hope that the PR could be merged in as soon as it has been fully tested. No one has ever reported that the current one is working or not, at least, it works on my testing VM. I hope that I could find all the bugs and make sure the scripts work as it should.

I'm not satisfied with the current ad-hoc WordPress solution. I strongly prefer that we use the official WordPress docker image as the base. Does it matter whether or not this one works if this isn't the approach we want to take?

**Comment by mattma** [comment]
will do.

**Comment by ericelliott** [comment]
Thanks. :smile:  :+1:

**Comment by mattma** [comment]
@ericelliott @robmazan 

It is using the offical `wordpress` image now. 

Read the configuration from user's setting.

Update the Readme. 

Please, follow the [instruction](https://github.com/mattma/JSHomes-Platform/blob/master/docs/README.md) to start your wordpress instance.

**Comment by ericelliott** [comment]
Thanks, will take a look asap!

**Comment by robmazan** [comment]
@mattma @ericelliott 

Successfully built it (after fixing line endings on Win). :+1: 

FYI, setting up `docker-machine` + Kitematic just got easier (plus `docker-compose` is included for Mac now by default): https://blog.docker.com/2015/08/docker-toolbox/

**Comment by mattma** [comment]
Great. @robmazan Does this version satisfied all of your previous comments? I know that you mentioned `Kitematic` approach. 

One of the requirement from @ericelliott is using `vagrant` and `virtualbox` provider. (development friendly way)

`Kitematic` is being a Docker Mac UI app (General public friendly), it is using `boot2docker` and `virtualbox` provider. It would be great to have two options in my opinion.

**Comment by robmazan** [comment]
@mattma 

I find this `.env` based `docker-compose.yml` generation a bit of an overkill, but this is most probably because I don't see the full picture. :smiley: 

From my previous comments one thing was not covered yet. Not a big deal, but I find it more stable to configure **private network** instead of port forwarding to avoid accidental port collisions. This might be again something, where I'm not aware of the reasons. Again, not some serious concern, just hasn't been discussed yet.

> Kitematic is being a Docker Mac UI app (General public friendly), it is using boot2docker and virtualbox provider. It would be great to have two options in my opinion.

Actually that's not hard at all. Just tried it. Running `docker-compose` from Kitematic's Docker CLI against the generated `docker-compose.yml` works smoothly. :smiley: 

One thing I've spotted since: some of the `Dockerfile` comments now became deprecated. I'm not even sure you need the *"# install the PHP extensions"*  section any more.

**Comment by mattma** [comment]
>  you need the "# install the PHP extensions" section any more.

You are right. This section is totally optional. I leave it there to update all the system packages and patches. `VIM` is a good tool to have if need to edit any files inside. `wget` is not needed.

> configure private network instead of port forwarding to avoid accidental port collisions

Sure. If user use the existed port, `vagrant` won't start due to conflict port. It is definitely not MVP. we should get this PR in ASAP, then address it in the future. 

> I find this .env based docker-compose.yml generation a bit of an overkill, but this is most probably because I don't see the full picture

`docker-compose.yml` is a static file, it cannot take any variable inside `yml` extension. How am I going to add `Environment` variable to the `image`? Another way that I can think of is, create a `Dockerfile` with `onbuild` optional, let user insert `Environment` in the build time. Well, that requires the image is built every time when wordpress starts. I think that is even more over-kill. 

Can you think of a better way? @robmazan

**Comment by robmazan** [comment]
Hi @mattma 

I just don't get the purpose of having that dynamic `docker-compose.yml` stuff. I mean why can't you say that the WordPress instance is called `wordpress`, the MySql instance is `db` the database name is `jshomes`, etc.?

This is most probably because I don't see all the goals you want to reach. Having `.env` file is only reasonable IMO, if you want to reuse the values defined in it in some other code (though even for example in PHP you can get the same values from the environment variables in the container).

Even if you need a different build configuration for each environment (like CI, TEST, PROD, etc), you can still create `docker-compose.ci.yml`, `docker-compose.test.yml`, etc. to cover that.

So again, as the complete usage scenario is not clear for me, I cannot tell if that part of the approach is good or not. I assume you have reasons behind doing things this way, it's just I don't know those reasons - that's why I was asking for some info about the context before. :smiley:

File: docs/README.md [hunk 1]
```diff
@@ -7,38 +7,66 @@ The JSHomes platform is a tech training courseware platform intended to support
...
+3. Edit configuration file:
 
-Edit `.env` file to configure your app instance:
+```bash
+cp .env.example .env
+vi .env
 
-```
-PF_BASE_URL=<base-url> # The URL it will be served from, i.e. `http://localhost`
+# Currently Not-in-use. it is not needed in Docker-compose orchestration
```
Comment (ericelliott) [inline docs/README.md]: If it's not needed, just delete it. =)

File: docs/README.md [hunk 2]
```diff
@@ -7,38 +7,66 @@ The JSHomes platform is a tech training courseware platform intended to support
...
-PF_LOGGED_IN_SALT=<your-value>
-PF_NONCE_SALT=<your-value>
+4. Start wordpress:
+
+```bash
+./start
 ```
 
-Set your site vhost document root to /path/to/site/web/ (/path/to/site/current/web/ if using deploys)
+5. View your `wordpress` in your favorate browser at `http://localhost:9888`:
```
Comment (ericelliott) [inline docs/README.md]: "favorate" should be "favorite"

**Comment by ericelliott** [comment]
I'm on a Mac. Not working for me.

```
vagrant@ubuntu-14:~/platform$ ./start
Stopping platform_localhost_1...
localhost uses an image, skipping
Building wordpress...
Step 0 : FROM wordpress
 ---> e04cc9e4c4c1
Step 1 : MAINTAINER Matt Ma "matt@mattmadesign.com"
 ---> Using cache
 ---> ec5e38ec47e0
Step 2 : ENV REFRESHED_AT 2015-08-12
 ---> Using cache
 ---> a153e033f6b8
Step 3 : RUN apt-get -qq update &&   apt-get -y upgrade &&   apt-get install -y vim wget &&   rm -rf /var/lib/apt/lists/*
 ---> Using cache
 ---> 6884dbabfb74
Step 4 : RUN rm -rf /usr/src/wordpress/wp-content/plugins/* &&   rm -rf /usr/src/wordpress/wp-content/themes/* &&   cp -fr /usr/src/wordpress/* /var/www/html/ &&   chown -R www-data:www-data /var/www/html/
 ---> Using cache
 ---> b362b05bd0b5
Step 5 : VOLUME /var/www/html/wp-content/plugins /var/www/html/wp-content/themes
 ---> Using cache
 ---> 972bc4e2b581
Step 6 : WORKDIR /var/www/html/
 ---> Using cache
 ---> ce74c520f6d9
Step 7 : EXPOSE 80 3306
 ---> Using cache
 ---> 6b0cbd774fa6
Step 8 : CMD apache2-foreground
 ---> Using cache
 ---> 9ed053b76cb8
Successfully built 9ed053b76cb8
Recreating platform_localhost_1...
Recreating platform_wordpress_1...
vagrant@ubuntu-14:~/platform$ netstat
Active Internet connections (w/o servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp        0    116 10.0.2.15:ssh           10.0.2.2:59735          ESTABLISHED
Active UNIX domain sockets (w/o servers)
Proto RefCnt Flags       Type       State         I-Node   Path
unix  7      [ ]         DGRAM                    8387     /dev/log
unix  3      [ ]         STREAM     CONNECTED     15140    /var/run/dbus/system_bus_socket
unix  3      [ ]         STREAM     CONNECTED     15137    
unix  3      [ ]         STREAM     CONNECTED     742      
unix  3      [ ]         STREAM     CONNECTED     829      
unix  3      [ ]         STREAM     CONNECTED     731      
unix  3      [ ]         STREAM     CONNECTED     730      
unix  3      [ ]         STREAM     CONNECTED     8338     
unix  3      [ ]         STREAM     CONNECTED     8427     @/com/ubuntu/upstart
unix  2      [ ]         DGRAM                    9368     
unix  3      [ ]         STREAM     CONNECTED     10195    /var/run/dbus/system_bus_socket
unix  3      [ ]         STREAM     CONNECTED     8337     
unix  2      [ ]         DGRAM                    12148    
unix  3      [ ]         STREAM     CONNECTED     10194    
unix  3      [ ]         STREAM     CONNECTED     547      @/com/ubuntu/upstart
unix  2      [ ]         DGRAM                    15080    
unix  3      [ ]         STREAM     CONNECTED     743      /var/run/dbus/system_bus_socket
unix  3      [ ]         STREAM     CONNECTED     543      
unix  3      [ ]         STREAM     CONNECTED     830      /var/run/dbus/system_bus_socket
unix  3      [ ]         STREAM     CONNECTED     15139    
unix  3      [ ]         STREAM     CONNECTED     15136    
unix  3      [ ]         STREAM     CONNECTED     756      @/com/ubuntu/upstart
unix  3      [ ]         STREAM     CONNECTED     708      
unix  3      [ ]         STREAM     CONNECTED     9222     
unix  3      [ ]         DGRAM                    7963     
unix  2      [ ]         DGRAM                    8872     
unix  3      [ ]         DGRAM                    7962     
unix  2      [ ]         DGRAM                    8589     
vagrant@ubuntu-14:~/platform$ 
```

```
vagrant@ubuntu-14:~/platform$ netstat --listen
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp        0      0 *:ssh                   *:*                     LISTEN     
tcp        0      0 *:49287                 *:*                     LISTEN     
tcp        0      0 *:sunrpc                *:*                     LISTEN     
tcp6       0      0 [::]:ssh                [::]:*                  LISTEN     
tcp6       0      0 [::]:49567              [::]:*                  LISTEN     
tcp6       0      0 [::]:sunrpc             [::]:*                  LISTEN     
udp        0      0 *:50720                 *:*                                
udp        0      0 *:691                   *:*                                
udp        0      0 localhost:730           *:*                                
udp        0      0 *:49149                 *:*                                
udp        0      0 *:bootpc                *:*                                
udp        0      0 *:sunrpc                *:*                                
udp        0      0 1.qarestr.sub-172-1:ntp *:*                                
udp        0      0 10.0.2.15:ntp           *:*                                
udp        0      0 localhost:ntp           *:*                                
udp        0      0 *:ntp                   *:*                                
udp6       0      0 [::]:691                [::]:*                             
udp6       0      0 [::]:11306              [::]:*                             
udp6       0      0 [::]:sunrpc             [::]:*                             
udp6       0      0 fe80::50d7:91ff:fe6:ntp [::]:*                             
udp6       0      0 fe80::5484:7aff:fef:ntp [::]:*                             
udp6       0      0 fe80::a00:27ff:fe98:ntp [::]:*                             
udp6       0      0 localhost:ntp           [::]:*                             
udp6       0      0 [::]:ntp                [::]:*                             
udp6       0      0 [::]:56450              [::]:*                             
Active UNIX domain sockets (only servers)
Proto RefCnt Flags       Type       State         I-Node   Path
unix  2      [ ACC ]     SEQPACKET  LISTENING     559      /run/udev/control
unix  2      [ ACC ]     STREAM     LISTENING     380      @/com/ubuntu/upstart
unix  2      [ ACC ]     STREAM     LISTENING     15806    /var/run/docker.sock
unix  2      [ ACC ]     STREAM     LISTENING     716      /var/run/dbus/system_bus_socket
unix  2      [ ACC ]     STREAM     LISTENING     8436     /run/rpcbind.sock
```

File: docs/README.md (hunk 1 above)
Comment (mattma) [inline docs/README.md]: fixed.

File: docs/README.md (hunk 2 above)
Comment (mattma) [inline docs/README.md]: fixed.

**Comment by mattma** [comment]
@ericelliott 

I am investigating the problem right now. I found an error

```
Warning: mysqli::mysqli(): (HY000/2002): No such file or directory in - on line 10

MySQL Connection Error: (2002) No such file or directory
```

Do not know why it works last week. It may be due to the latest release of wordpress image.

**Comment by ericelliott** [comment]
Thanks, @mattma - There's a problem with the theme, but we'll open a new issue to address that.

**Comment by mattma** [comment]
@ericelliott Great. I am very happy to see this PR finally get merged. Hopefully, it unblocks the pipeline.

**Comment by ericelliott** [comment]
Yep. I'm excited to get this thing rolling. EricElliottJS.com has been in production for months, and I've started copying a lot of what we're using there into our todo list here. If you want to knock out some quick wins, there are lots of issues that are just about adding WordPress plugins or tweaking WordPress settings. You'll find them tagged [WordPress](https://github.com/jshomes/JSHomes-Platform/labels/wordpress).

[This issue](https://github.com/jshomes/JSHomes-Platform/issues/20) looks like you could knock it out quickly. Are you game?

**Comment by mattma** [comment]
I will give it a try when I have free time.

**Pull request merged**